### ISSUE TITLE
Wheres the docs on slate script files, vendors etc.

### ISSUE THREAD

**Issue by joshbedo** [Shopify/slate#377]
I have a section that contains the following code

```
<div data-section-id="{{ section.id }}" data-section-type="instagram-section">
  <div class="container">
    <div class="row">
      <div class="instafeed-container">
        <div class="instafeed-content">
          <h2>Instafeed</h2>
          <div class="instafeed-desc">Some description</div>
        </div>
        <div id="instafeed" class="instafeed">
        </div>
      </div>
    </div>
  </div>
</div>
<script type="text/javascript">
  var feed = new Instafeed({
    get: 'user',
    userId: {{ section.settings.instagram_id }},
    accessToken: '{{ section.settings.instagram_access_token }}',
    after: function() {
      $('#instafeed a > img').each(function(i) {
        var src = $(this).attr('src').replace('/s150x150/', '/s320x320/');
        $(this).attr('src', src);
      });
    },
    limit: {{ section.settings.limit }}
  });
</script>
{% schema %}
{
  "name": "Instagram",
  "class": "index-section-instagram",
  "settings": [
    {
      "type": "text",
      "id": "title",
      "label": "Title"
    },
    {
      "type": "textarea",
      "id": "description",
      "label": "Description (HTML)"
    },
    {
      "type": "text",
      "id":"instagram_id",
      "label":" Instagram User ID",
      "info": "This is the Instagram User ID, not the username"
    },
    {
      "type": "text",
      "id": "instagram_access_token",
      "label":"Instagram Access Token"
    },
    {

      "type": "text",
      "id":"limit",
      "label":"Limit images"
    }
  ],
    "presets": [{
    "name": "Instagram",
    "category": "Store Information"
  }]
}
{% endschema %}
```

but the script breaks because no dependencies have loaded. How can i fix this to make sure vendors are loaded before the section JS tries to execute.

**Comment by ADTC** [comment]
This happens because there is a [`defer` attribute](https://davidwalsh.name/script-defer) on the script tag in `theme.liquid` which loads the `vendor.js` script. This is a script that encloses jQuery and Modernizr by default.
```
  <!--[if (gt IE 9)|!(IE)]><!--><script src="{{ 'vendor.js' | asset_url }}" defer="defer"></script><!--<![endif]-->
  <!--[if lt IE 9]><script src="{{ 'vendor.js' | asset_url }}"></script><![endif]-->
```
Remove the `defer` attribute. Actually, when you remove it from the code above, both `if` blocks end up with the same script tag. Since the `if` blocks semantically function like `if-else`, they can be removed and the code can be simplified to this instead:
```
  <script src="{{ 'vendor.js' | asset_url }}"></script>
```
Also, **it is recommended (but not required)** to put your code in a `ready` block:
```
$(document).ready(function() {
    // your code here
});
```
**PS:** You may also need to fix this in `gift_card.liquid` and `password.liquid` depending on how you're building your theme.

**Comment by ADTC** [comment]
I found out that this happens because the jQuery (and Modernizr) is loaded through `vendor.js` using `defer` attribute, which delays the loading of jQuery, hence breaking the dependency of your script.

Now, you could fix this easily by removing the `defer` attribute, which was my original suggestion. But doing it this way would completely break the Theme Editor (that thing a user sees when they click Customize on the theme in Shopify Admin). This is because, the Theme Editor reloads the section on changing its settings - it doesn't reload the whole page or run that `<script>` again. You have to listen to events and manually trigger JavaScript when the section reload happens.

So after some trial and error, I finally have the proper way to do this (there are several things that need to change):

1. Change your script tag to use Liquid javascript directive instead. So, this:
```html
<script type="text/javascript">
...
</script>
```
becomes this:
```php
{% javascript %}
...
{% endjavascript %}
```
**What this does:** Liquid parser will put that into a deferred script file. This goes into one of the scripts injected using `content_for_header` in `theme.liquid`. Since this is loaded after the `defer`-loaded jQuery, it will work fine and won't break because of missing jQuery.

**Caveat:** The `javascript` directive is a special directive similar to the `schema` directive. You can only put one at the end of your section Liquid code (before `schema`). So you cannot put it in between your HTML code or base it on a condition.

**Tip:** If you want the syntax highlighting of your IDE to apply on the JS code, you may be able to use this instead (the commented script tags act as hints to IDE but otherwise have no effect):
```php
{% javascript %}
//<script>
...
//</script>
{% endjavascript %}
```
2. Next, pump your section settings into data attributes. This could be any HTML tag in your HTML code, but for simplicity you could use the top-most tag or a uniquely named tag (which you may use anyway). For example:
```html
<div class="instafeed-container">
```
becomes
```html
<div class="instafeed-container" data-settings='{
"instagram_id": {{ section.settings.instagram_id }},
"instagram_access_token": "{{ section.settings.instagram_access_token }}",
"limit": {{ section.settings.limit }}
}'>
<!--
Note:
"instagram_id": {{ section.settings.instagram_id }}, // numeric (no quotes)
"instagram_access_token": "{{ section.settings.instagram_access_token }}", // string in quotes
"limit": {{ section.settings.limit }} // no comma here
-->
```
3. Access the data attribute in your Javascript and get the settings from there. So this:
```js
userId: {{ section.settings.instagram_id }},
accessToken: '{{ section.settings.instagram_access_token }}',
...
limit: {{ section.settings.limit }}
```
becomes this:
```js
var settings = $('.instafeed-container').data('settings');
...
userId: settings.instagram_id,
accessToken: settings.instagram_access_token,
...
limit: settings.limit
```
4. [Read this](https://help.shopify.com/themes/development/sections#understand-the-interaction-between-theme-javascript-and-the-theme-editor). And then, enclose up your Javascript so that Theme Editor can work properly to re-execute it upon section reload:
```js
{% javascript %}

function my_load() {
  // JS goes here
}
function my_unload() {
  // you will want to do clean-up and/or destroy what you created in my_load
}
function my_event_handler(event) {
  // there are cleaner ways to write this, but the below works for me
  if (event.detail.sectionId == 'my-section-name') {
    if (event.type == 'shopify:section:load') {
      my_load();
    } else if (event.type == 'shopify:section:unload') {
      my_unload();
    }
  }
}

$(document).ready(function() {
  my_load(); // this is to execute on initial page load
  $(document).on('shopify:section:load shopify:section:unload', my_event_handler);
});

{% endjavascript %}
```
_**Note: Use unique section-specific function names in the code above. The above is just an example.**_

5. TEST! Not only do you need to test that the section works fine when you load the page initially, you also need to test whether changing the section settings in Theme Customize (Theme Editor) will correctly reload the section with the new settings.

---
> _Where's your last comment? (in case you got an email notification but can't find it here anymore)_

I deleted it because it contains misinformation. This comment is a replacement for it, and a better, proper solution to the problem.

**Comment by t-kelly** [comment]
This is not a bug with Slate. Closing.

**Comment by ADTC** [comment]
True. But perhaps it can help improve documentation. :)

**Comment by t-kelly** [comment]
Not sure if documentation on how the effect's of `<script defer>` belong in Slate? I feel like that's a more general Front-End topic.

**Comment by ADTC** [comment]
Yes, I agree the effects of `defer` would be a general front-end topic. But my solution on how to fix the problem is very specific to Shopify and Liquid. The general solution would be to always use `defer` or don't use it at all, but neither of them seemed to work with the unique way Shopify behaves.

I don't expect it to be included in any basic documentation, but I'd like to see it as part of a troubleshooting guide. Maybe for the question _"Why can't I use jQuery in my `<script>` tag?"_

**Comment by sebbean** [comment]
running into issues with this as well.
trying to run some simple jquery in a script is at first glance

@ADTC i'm getting  `Liquid syntax error (line 7): Unknown tag 'javascript'` when using that your method (albeit inside a snippets file)
